File: src/java/_2391_Minimum_Amount_of_Time_to_Collect_Garbage.java
class Solution {
    public int garbageCollection(String[] garbage, int[] travel) {
        int n = garbage.length, ans = 0;

        for(int time: travel)
            ans += 3 * time;

        for(String house: garbage)
            ans += house.length();

        for(int i = n - 1; i > 0; i--)
            if (!garbage[i].contains("G"))
                ans -= travel[i - 1];
            else
                break;

        for(int i = n - 1; i > 0; i--)
            if (!garbage[i].contains("P"))
                ans -= travel[i - 1];
            else
                break;

        for(int i = n - 1; i > 0; i--)
            if (!garbage[i].contains("M"))
                ans -= travel[i - 1];
            else
                break;

        return ans;
    }
}

File: src/python/_2391_Minimum_Amount_of_Time_to_Collect_Garbage.py
class Solution:
    def garbageCollection(self, garbage: List[str], travel: List[int]) -> int:
        prefix = [0] + list(accumulate(travel))

        ans = len("".join(garbage))
        xIdx = {"M": 0, "G": 0, "P": 0}
        i = len(garbage) - 1

        while i >= 0 and xIdx["M"] * xIdx["G"] * xIdx["P"] == 0:
            for c in "MGP":
                if c in garbage[i]:
                    xIdx[c] = max(xIdx[c], i)

            i -= 1

        for idx in xIdx.values():
            ans += prefix[idx]

        return ans
